service/service.go

```diff
@@ -279,168 +328,145 @@ func (s *Service) Started() bool {
+	} else {
+		logger.With().Warning("execution failed", log.Err(err))
 	}
 
-	log.Info("Round %v execution ended, phi=%x, duration %v", r.ID, r.execution.NIP.Root, time.Since(start))
-
-	return nil
+	return roundResult{round: r, err: err}
 }
 
-func (s *Service) Submit(ctx context.Context, challenge, signature []byte) (*round, []byte, error) {
-	logger := logging.FromContext(ctx)
-	if !s.Started() {
-		return nil, nil, ErrNotStarted
+func (s *ServiceClient) Submit(ctx context.Context, challenge, signature []byte) (string, []byte, error) {
+	if s == nil {
+		return "", nil, ErrNotStarted
 	}
+	logger := logging.FromContext(ctx)
 
 	logger.Debug("Received challenge")
 	// SAFETY: it will never panic as `s.ChallengeVerifier` is set in Start
 	verifier := s.challengeVerifier.Load().(challenge_verifier.Verifier)
 	result, err := verifier.Verify(ctx, challenge, signature)
 	if err != nil {
 		logger.With().Debug("challenge verification failed", log.Err(err))
-		return nil, nil, err
+		return "", nil, err
 	}
 	logger.With().Debug("verified challenge", log.String("hash", hex.EncodeToString(result.Hash)), log.String("node_id", hex.EncodeToString(result.NodeId)))
-	s.openRoundMutex.Lock()
-	r := s.openRound
-	err = r.submit(result.NodeId, result.Hash)
-	s.openRoundMutex.Unlock()
+
+	type response struct {
+		round string
+		err   error
+	}
+	done := make(chan response, 1)
+	s.command <- func(s *Service) {
+		done <- response{
+			round: s.openRound.ID,
+			err:   s.openRound.submit(result.NodeId, result.Hash),
+		}
+		close(done)
+	}
+	resp := <-done
+
 	switch {
-	case errors.Is(err, ErrChallengeAlreadySubmitted):
-		return r, result.Hash, nil
+	case errors.Is(resp.err, ErrChallengeAlreadySubmitted):
+		return resp.round, result.Hash, nil
 	case err != nil:
-		return nil, nil, err
+		return "", nil, resp.err
 	}
-	return r, result.Hash, nil
+	logger.With().Debug("submitted challenge for round", log.String("round", resp.round))
+	return resp.round, result.Hash, nil
 }
 
-func (s *Service) Info() (*InfoResponse, error) {
-	if !s.Started() {
+func (s *ServiceClient) Info() (*InfoResponse, error) {
+	if s == nil {
 		return nil, ErrNotStarted
 	}
 
-	s.executingRoundsMutex.RLock()
-	ids := make([]string, 0, len(s.executingRounds))
-	for id := range s.executingRounds {
-		ids = append(ids, id)
+	resp := make(chan *InfoResponse, 1)
+	s.command <- func(s *Service) {
+		ids := make([]string, 0, len(s.executingRounds))
+		for id := range s.executingRounds {
+			ids = append(ids, id)
+		}
+		resp <- &InfoResponse{
+			OpenRoundID:        s.openRound.ID,
+			ExecutingRoundsIds: ids,
+		}
+		close(resp)
 	}
-	s.executingRoundsMutex.RUnlock()
-
-	return &InfoResponse{
-		OpenRoundID:        s.openRoundID(),
-		ExecutingRoundsIds: ids,
-	}, nil
+	return <-resp, nil
 }
 
 // newRound creates a new round with the given epoch. This method MUST be guarded by a write lock on openRoundMutex.
```

Comment:
I feel like this comment isn't accurate any more.

Also, why does this function `panic`?

Reply:
Will update the doc, thanks. It panics if it couldn't save the state (this part is unchanged in the PR).

Reply:
So if the state fails to persist poet just crashes silently?

Reply:
Why silently? It should emit a backtrace.

Reply:
Silently in a sense as it doesn't even log why it decided to panic.

I know there will be a backtrace but that is not a structured log message that cannot easily be searched for e.g. in 10_000 of logs that we emit during a systest.

Reply:
A panic prints the message of the error that is passed in. The error should already contain the _why_ and _what_.

Reply:
The error that the panic prints will be something along the lines of `"permission denied"` or `"file already closed"` without any further context

Reply:
Resolving as the `newRound` documentation is fixed. 

@mafa The panic topic can be continued. Please create an issue if this needs to be addressed in your opinion.